doc: nrf_modem: add documentation for NRF_SO_KEEPOPEN (#40)
lemrey/sdk-nrfxlib: nrf_modem/doc/sockets/socket_options_func.rst

+-----------------+---------------------------------+------------------------+------------+--------------------------------------------------------------------------------------------+
| NRF_SOL_SOCKET  | NRF_SO_EXCEPTIONAL_DATA         | ``int``                | get/set    | Send data on socket as part of exceptional event.                                          |
+-----------------+---------------------------------+------------------------+------------+--------------------------------------------------------------------------------------------+

> | NRF_SOL_SOCKET  | NRF_SO_KEEPOPEN                 | ``int``                | get/set    | Keep the socket open when its PDN connection is lost, or the device is put in flight mode. |

Reviewer: Remind me again, do we get the PDN down event in case of CFUN=4? If we lose the PDN connection in flight mode, maybe we can use flight mode as an example e.g. when the device is put in flight mode or out of reach from the base station. What about CFUN=20 (deactivate LTE)?

Author: Basically the application can `CFUN=4` and keep the socket open . It is not the case with `CFUN=0`.
`CFUN=20` seems to work the same as `CFUN=4`.

> | NRF_SOL_SOCKET  | NRF_SO_KEEPOPEN                 | ``int``                | get/set    | Keep the socket open when its PDN connection is lost, or the device is put in flight mode. |

Author: This changes the meaning of the sentence, what I mean to say is that the socket will remain open when the PDN connection is lost, or in case the device is put in flight mode (which disables connectivity).

> | NRF_SOL_SOCKET  | NRF_SO_KEEPOPEN                 | ``int``                | get/set    | Keep the socket open when its PDN connection is lost, or the device is put in flight mode. |

Reviewer 2: okay, my bad, I misunderstood that the socket option does two actions here.
Then,  we could rephrase as : _Keep the socket open when its PDN connection is lost or the device is set to flight mode (``CFUN=4``)._

> | NRF_SOL_SOCKET  | NRF_SO_KEEPOPEN                 | ``int``                | get/set    | Keep the socket open when its PDN connection is lost, or the device is put in flight mode. |

Author: Ok, I rephrased it but kept the `CFUN=4` out of here since it's only a short summary in a table; we mention in it in the description below.

   The socket option is supported from modem firmware v2.0.0.

NRF_SO_KEEPOPEN

>    Keep the socket from forcibly being closed upon PDN disconnection or reactivation events, or when the device is put into flight mode.

Reviewer: ```suggestion
   Keep the socket from being closed upon PDN disconnection or reactivation events, or when the device is put into flight mode.
```

Reviewer: Do we need to separate flight mode?

Author: I think we should mention that `CFUN=4` let the socket remain open (if possible).

Reviewer: ```suggestion
   Keep the socket from forcibly being closed upon PDN disconnection or reactivation events, or when the device is put into flight mode (CFUN=4).
```

>    Keep the socket from forcibly being closed upon PDN disconnection or reactivation events, or when the device is put into flight mode.

Reviewer 2: ```suggestion
   Keep the socket from being closed upon PDN disconnection or reactivation events, or when the device is put into flight mode  (``CFUN=4``).
```
Combined Evind's suggestions with a minor suggestion from my side.

Author: Changed

>    Until the PDN connection is reestablished, the socket is not functional and socket operations will return errors as appropriate (in general, ``NRF_ENETUNREACH``).

Reviewer 3: ```suggestion
   Until the PDN connection is reestablished, the socket is not functional and socket operations will return errors as appropriate (in general, :c:macro:`NRF_ENETUNREACH`).
```

Author: Updated

>    Until the PDN connection is reestablished, the socket is not functional and socket operations will return errors as appropriate (in general, ``NRF_ENETUNREACH``).

Reviewer 2: ```suggestion
   Until the PDN connection is reestablished, the socket is not functional, and socket operations return errors as appropriate (in general, :c:macro:`NRF_ENETUNREACH`).
```

Author: Updated

>    Upon PDN connection reestablishment, the socket behavior depends on the socket type and protocol, and on the IP address of the socket's newly established PDN connection, as shown in the table below.

Reviewer 2: ```suggestion
   Upon PDN connection reestablishment, the socket behavior depends on the socket type and protocol and on the IP address of the socket's newly established PDN connection, as shown in the following table:
```

Author: Updated

   +-----------------+--------------------+--------------------------------------------------------------+
   | Socket type     | Socket protocol    | Socket is functional (no errors)                             |
   +=================+====================+==============================================================+
   | NRF_SOCK_DGRAM  | NRF_IPPROTO_UDP    | Always                                                       |
   +-----------------+--------------------+--------------------------------------------------------------+

>    | NRF_SOCK_DGRAM  | NRF_SPROTO_DTLS1v2 | If peer supports DTLS connection ID                          |

Reviewer: ```suggestion
   | NRF_SOCK_DGRAM  | NRF_SPROTO_DTLS1v2 | If peer supports DTLS connection ID and this is set on the socket      |
```

   +-----------------+--------------------+--------------------------------------------------------------+
   | NRF_SOCK_STREAM | NRF_IPPROTO_TCP    | If the IP address of socket's PDN connection has not changed |
   +-----------------+--------------------+--------------------------------------------------------------+
   | NRF_SOCK_STREAM | NRF_SPROTO_TLS1v2  | If the IP address of socket's PDN connection has not changed |
   +-----------------+--------------------+--------------------------------------------------------------+
   | NRF_SOCK_RAW    | Any                | Always                                                       |
   +-----------------+--------------------+--------------------------------------------------------------+

>    If the conditions to keep the socket open after PDN connection reestablishment are not met, the socket will report an error (in general, ``NRF_ENETDOWN``), and must be closed by the application.

Reviewer 3: ```suggestion
   If the conditions to keep the socket open after PDN connection reestablishment are not met, the socket will report an error (in general, :c:macro:`NRF_ENETDOWN`), and must be closed by the application.
```

Author: OK

>    If the conditions to keep the socket open after PDN connection reestablishment are not met, the socket will report an error (in general, ``NRF_ENETDOWN``), and must be closed by the application.

Reviewer 2: ```suggestion
   If the conditions to keep the socket open after PDN connection reestablishment are not met, the socket reports an error (in general, ``NRF_ENETDOWN``), and must be closed by the application.
```

Author: Updated

   For further information on how to detect and handle these errors, see :ref:`handling_pdn_errors_on_sockets`.

.. note::
   Putting the device into functional mode ``0`` (``+CFUN=0``) always forces all sockets to be closed, regardless of the :c:macro:`NRF_SO_KEEPOPEN` socket option.

>    This socket option is supported from modem firmware v2.0.1.

Reviewer 2: This could be added to the public doc once the modem firmware v2.0.1 is released?

   The socket option is supported from modem firmware v2.0.0.

NRF_SO_KEEPOPEN

>    Keep the socket from being closed upon PDN disconnection or reactivation events, or when the device is put into flight mode (``+CFUN=4``).

Author: Thank you

   Until the PDN connection is reestablished, the socket is not functional.

>    Output operations such as :c:func:`nrf_send` and :c:func:`nrf_connect` return an error and set ``errno`` to :c:macro:`NRF_ENETUNREACH`.
>    Input operations such as :c:func:`nrf_recv` and :c:func:`nrf_accept` block, since no data can be received, or return no data (if the socket or the operation is non-blocking).

Reviewer 2: ```suggestion
   Output operations, such as the functions :c:func:`nrf_send` and :c:func:`nrf_connect` return an error and set ``errno`` to :c:macro:`NRF_ENETUNREACH`.
   Input operations, such as the functions :c:func:`nrf_recv` and :c:func:`nrf_accept` block, since no data can be received, or return no data (if the socket or the operation is non-blocking).
```

Author: Updated

   +-----------------+--------------------+--------------------------------------------------------------+
   | NRF_SOCK_DGRAM  | NRF_SPROTO_DTLS1v2 | If using DTLS connection ID                                  |
   +-----------------+--------------------+--------------------------------------------------------------+
   | NRF_SOCK_STREAM | NRF_IPPROTO_TCP    | If the IP address of socket's PDN connection has not changed |
   +-----------------+--------------------+--------------------------------------------------------------+
   | NRF_SOCK_STREAM | NRF_SPROTO_TLS1v2  | If the IP address of socket's PDN connection has not changed |
   +-----------------+--------------------+--------------------------------------------------------------+
   | NRF_SOCK_RAW    | Any                | Always                                                       |
   +-----------------+--------------------+--------------------------------------------------------------+

   If the conditions to keep the socket open after PDN connection reestablishment are not met, the socket will report an error (and set `errno` to :c:macro:`NRF_ENETDOWN`), and must be closed by the application.

>    Otherwise, the socket is functional and the application does not need to close it.

Reviewer 2: ```suggestion
   Otherwise, the socket is functional, and the application does not need to close it.
```

Author: Added a comma
